pkg/util/informermanager/single-cluster-manager.go

```diff
@@ -44,7 +44,7 @@ type SingleClusterInformerManager interface {
 	Stop()
 
 	// WaitForCacheSync waits for all caches to populate.
-	WaitForCacheSync() map[schema.GroupVersionResource]bool
+	WaitForCacheSync(stopCh <-chan struct{}) map[schema.GroupVersionResource]bool
```

Comment:
I suggest adding a new function `WaitForCacheSyncWithTimeout` here instead of modifying `WaitForCacheSync`, so you don't have to pass the `singleClusterInformerManager.Context()`. The `WaitForCacheSync` does not need a stop channel because we had a context.

```go
func (s *singleClusterInformerManagerImpl) WaitForCacheSyncWithTimeout(cacheSyncTimeout time.Duration) map[schema.GroupVersionResource]bool {
	ctx, cancel := context.WithTimeout(s.ctx, cacheSyncTimeout)
	defer cancel()
	s.lock.Lock()
	defer s.lock.Unlock()
	res := s.informerFactory.WaitForCacheSync(ctx.Done())
	for resource, synced := range res {
		if _, exist := s.syncedInformers[resource]; !exist && synced {
			s.syncedInformers[resource] = struct{}{}
		}
	}
	return res
}
```

Reply:
+1
In addition, we can add an internal shared method like `waitForCacheSync`:
```golang
func (s *singleClusterInformerManagerImpl) WaitForCacheSync() map[schema.GroupVersionResource]bool {
	s.lock.Lock()
	defer s.lock.Unlock()
	return s.waitForCacheSync(s.ctx)
}

func (s *singleClusterInformerManagerImpl) WaitForCacheSyncWithTimeout(cacheSyncTimeout time.Duration) map[schema.GroupVersionResource]bool {
	ctx, cancel := context.WithTimeout(s.ctx, cacheSyncTimeout)
	defer cancel()
	s.lock.Lock()
	defer s.lock.Unlock()

	return s.waitForCacheSync(ctx)
}

func (s *singleClusterInformerManagerImpl) waitForCacheSync(ctx context.Context) map[schema.GroupVersionResource]bool {
	res := s.informerFactory.WaitForCacheSync(ctx.Done())
	for resource, synced := range res {
		if _, exist := s.syncedInformers[resource]; !exist && synced {
			s.syncedInformers[resource] = struct{}{}
		}
	}
	return res
}
```